### PR TITLE
trie: polish committer

### PR DESCRIPTION
**Pull request by rjl493456442**
This RP contains some polish for trie committer.

### CODE REVIEW

File: trie/committer.go [hunk 1]
```diff
@@ -104,87 +97,93 @@ func (c *committer) commit(n node, db *Database, force bool) (node, error) {
 	case *shortNode:
 		// Commit child
 		collapsed := cn.copy()
-		if _, ok := cn.Val.(valueNode); !ok {
-			childV, err := c.commit(cn.Val, db, false)
+
+		// If the child is fullnode, recursively commit.
+		// Otherwise it can only be hashNode or valueNode.
+		if _, ok := cn.Val.(*fullNode); ok {
```
Comment (holiman) [inline trie/committer.go:103]: Can't the child be another shortnode? 
Let's say you have a small trie. The root will be a shortnode, when you start inserting, you will have only shortnodes until they start getting crowded, no?

File: trie/committer.go [hunk 2]
```diff
@@ -104,87 +97,93 @@ func (c *committer) commit(n node, db *Database, force bool) (node, error) {
...
-				return n // Nodes smaller than 32 bytes are stored inside their parent
-			}
-			hash = c.makeHashNode(c.tmp)
-		} else {
-			// This was not generated - must be a small node stored in the parent
-			// No need to do anything here
-			return n
-		}
+		// This was not generated - must be a small node stored in the parent.
+		// In thoery we should apply the leafCall here if it's not nil(embeded
```
Comment (holiman) [inline trie/committer.go:183]:
```suggestion
		// In theory we should apply the leafCall here if it's not nil(embeded
```

File: trie/committer.go [hunk 3]
```diff
@@ -104,87 +97,93 @@ func (c *committer) commit(n node, db *Database, force bool) (node, error) {
...
-			if size < 32 && !force {
-				return n // Nodes smaller than 32 bytes are stored inside their parent
-			}
-			hash = c.makeHashNode(c.tmp)
-		} else {
-			// This was not generated - must be a small node stored in the parent
-			// No need to do anything here
-			return n
-		}
+		// This was not generated - must be a small node stored in the parent.
```
Comment (holiman) [inline trie/committer.go:182]: This is a substantial change. I don't understand why this is correct. Can you elaborate?

File: trie/committer.go (hunk 1 above)
Comment (rjl493456442) [reply trie/committer.go:103]: The child of a shortnode can't be the shortnode. If so, they can be compressed into a single one. Look at here https://github.com/ethereum/go-ethereum/blob/master/trie/trie.go#L299

> The root will be a shortnode, when you start inserting, you will have only shortnodes until they start getting crowded

When you insert some entries, the shortnode will be converted into a fullnode, the inserted entry will be created as the shortnode.

File: trie/committer.go (hunk 3 above)
Comment (rjl493456442) [reply trie/committer.go:182]: When we try to commit the trie, we will first hash the trie. This assumption can always be held.
So if the cached hash is missing, there is only one possibility: the node is too small to have the hash.

The code path I delete should never be triggered. We never calculate the hash for value directly. 
We only calculate the hash for "shortNode" + "fullNode". The value is always embedded in the shortnode.

**Comment by rjl493456442** [comment]
@holiman Thanks for reviewing. The PR contains a few facts(I think) for some cleanup.

* The fullnode should never have valuenode child in the range [0, 15]. 
All value nodes(except the 17th) are wrapped by a shortnode https://github.com/ethereum/go-ethereum/blob/master/trie/trie.go#L236
* The shortnode will never have a direct shortnode child. If so these two will be combined into a single one.
https://github.com/ethereum/go-ethereum/blob/master/trie/trie.go#L299
* The committer will commit the trie only when the trie is already hashed

These facts to be honest is sensitive, would be nice to be confirmed from @fjl and @karalabe

**Review by holiman — APPROVED** [review]
LGTM. @rjl493456442  -- want to un-draft this?

**Comment by rjl493456442** [comment]
@holiman Maybe I should close this PR. Because I had another big PR(background committer) based on this.

Re the background committer: https://github.com/ethereum/go-ethereum/pull/21377

**Comment by holiman** [comment]
> Because I had another big PR(background committer) based on this. 

Well, this one is fairly small and nicely self-contained, whereas the other one is a pretty large refactor... Might be a good approach to do this one first, if the other one incorporates these changes too.

**Comment by rjl493456442** [comment]
@holiman Yes, I rebased this one and let's merge this tiny one first.

**Comment by rjl493456442** [comment]
@holiman I removed the `hasValue` notion in trie committer.

According to the benchmarks, type assertion between the pointer and
interface is extremely fast.
    
    BenchmarkIntmethod-12           1000000000               1.91 ns/op
    BenchmarkInterface-12           1000000000               2.13 ns/op
    BenchmarkTypeSwitch-12          1000000000               1.81 ns/op
    BenchmarkTypeAssertion-12       2000000000               1.78 ns/op
    
So the overhead for asserting whether the shortnode has "valuenode"
child is super tiny. No necessary to have another field.

**Comment by holiman** [comment]
> So the overhead for asserting whether the shortnode has "valuenode"
child is super tiny. No necessary to have another field.

Yes, it made more sense earlier, when we iterated over all elements -- but you already optimized that away :)

File: trie/committer.go (hunk 1 above)
Comment (rjlrjl) [reply trie/committer.go:102]: Hi,
Somehow I ended up on this git email distribution list.  Can somebody please remove me??  I get emails three times a day with this subject:  

[ethereum/go-ethereum] trie: polish committer (#21351)

Comment (rjl493456442) [reply trie/committer.go:102]: Sorry, removed.

**Comment by holiman** [comment]
It's really odd -- I don't understand why I'm not seeing any improvements here...? 
```
name                             old time/op    new time/op    delta
CommitAfterHash/no-onleaf-6        3.36µs ±23%    2.99µs ±34%    ~     (p=0.310 n=5+5)
CommitAfterHash/with-onleaf-6      3.46µs ±31%    3.54µs ±18%    ~     (p=0.841 n=5+5)
CommitAfterHashFixedSize/10-6      29.3µs ±10%    29.0µs ± 1%    ~     (p=0.690 n=5+5)
CommitAfterHashFixedSize/100-6      133µs ± 3%     136µs ± 1%  +2.41%  (p=0.032 n=5+5)
CommitAfterHashFixedSize/1K-6      1.35ms ± 4%    1.37ms ± 3%    ~     (p=0.690 n=5+5)
CommitAfterHashFixedSize/10K-6     21.2ms ± 5%    21.9ms ± 4%    ~     (p=0.421 n=5+5)
CommitAfterHashFixedSize/100K-6     262ms ± 7%     282ms ± 9%    ~     (p=0.056 n=5+5)

name                             old alloc/op   new alloc/op   delta
CommitAfterHash/no-onleaf-6          792B ± 9%      744B ± 1%    ~     (p=0.667 n=5+5)
CommitAfterHash/with-onleaf-6      1.03kB ± 4%    1.02kB ± 3%    ~     (p=0.841 n=5+5)
CommitAfterHashFixedSize/10-6      17.0kB ± 0%    17.0kB ± 0%    ~     (p=0.460 n=5+5)
CommitAfterHashFixedSize/100-6     80.6kB ± 0%    80.6kB ± 0%    ~     (p=0.881 n=5+5)
CommitAfterHashFixedSize/1K-6       796kB ± 0%     796kB ± 0%    ~     (p=0.841 n=5+5)
CommitAfterHashFixedSize/10K-6     9.28MB ± 0%    9.28MB ± 0%    ~     (p=0.310 n=5+5)
CommitAfterHashFixedSize/100K-6    86.0MB ± 0%    86.0MB ± 0%    ~     (p=0.841 n=5+5)

name                             old allocs/op  new allocs/op  delta
CommitAfterHash/no-onleaf-6          7.00 ± 0%      7.00 ± 0%    ~     (all equal)
CommitAfterHash/with-onleaf-6        13.0 ± 0%      13.0 ± 0%    ~     (all equal)
CommitAfterHashFixedSize/10-6         162 ± 0%       162 ± 0%    ~     (all equal)
CommitAfterHashFixedSize/100-6        768 ± 0%       768 ± 0%    ~     (all equal)
CommitAfterHashFixedSize/1K-6       7.83k ± 0%     7.83k ± 0%    ~     (all equal)
CommitAfterHashFixedSize/10K-6      80.1k ± 0%     80.1k ± 0%    ~     (p=0.389 n=5+5)
CommitAfterHashFixedSize/100K-6      798k ± 0%      798k ± 0%    ~     (p=0.841 n=5+5)
```

**Comment by rjl493456442** [comment]
Well... Why do we can get any improvement here?
For leaf callback, originally we had a flag to skip most of the checkings.

**Pull request merged**